misk-aws/src/main/kotlin/misk/jobqueue/sqs/SqsJobQueue.kt

```diff
@@ -22,82 +26,151 @@ internal class SqsJobQueue @Inject internal constructor(
-                  .withDataType("String")
-                  .withStringValue(value)
-              )
-            }
+    executeWithTracingAndErrorHandling(queueName, jobs.size) { span: Span, queue: ResolvedQueue ->
+      val (sendDuration, batchResult) = queue.call { client ->
+          val messageEntries = jobs.map { message ->
+            checkAttributeSize(message.attributes)
 
-            val metadata = mutableMapOf(
-              SqsJob.JOBQUEUE_METADATA_ORIGIN_QUEUE to queueName.parentQueue.value,
-              SqsJob.JOBQUEUE_METADATA_IDEMPOTENCE_KEY to idempotenceKey
-            )
+            SendMessageBatchRequestEntry(
+              message.idempotenceKey,
+              message.body,
+            ).apply {
+              if (message.deliveryDelay != null) delaySeconds =
+                getDelaySeconds(message.deliveryDelay)
+              message.attributes.forEach { (key, value) ->
+                addMessageAttributesEntry(
+                  key,
+                  value.toMessageAttributeValue()
+                )
+              }
 
-            // Preserve original trace id, if available.
-            (span as? DDSpan)?.let {
-              val traceId = it.context().traceId.toString()
-              metadata[SqsJob.JOBQUEUE_METADATA_ORIGINAL_TRACE_ID] = traceId
-              // TODO(bruno): drop this attribute after rollout; moved to metadata
               addMessageAttributesEntry(
-                SqsJob.ORIGINAL_TRACE_ID_ATTR,
-                MessageAttributeValue().withDataType("String").withStringValue(traceId)
+                SqsJob.JOBQUEUE_METADATA_ATTR,
+                createMetadataMessageAttributeValue(span, queueName, message.idempotenceKey)
               )
             }
-
-            // Add the internal metadata dictionary, encoded as JSON.
-            addMessageAttributesEntry(
-              SqsJob.JOBQUEUE_METADATA_ATTR,
-              MessageAttributeValue()
-                .withDataType("String")
-                .withStringValue(moshi.adapter<Map<String, String>>().toJson(metadata))
-            )
           }
 
-          timed { client.sendMessage(sendRequest) }
+          timed {
+            client.sendMessageBatch(SendMessageBatchRequest(queue.url, messageEntries))
+          }
         }
 
+      if(batchResult.failed.size > 0) {
+        throw JobQueue.BatchEnqueueException(queueName, batchResult.successful.map { it.id }, batchResult.failed.map {
+          JobQueue.EnqueueErrorResult(
+            it.id,
+            it.isSenderFault,
+            it.code,
+            it.message
+          )
+        })
+      }
+
+      return@executeWithTracingAndErrorHandling sendDuration
+    }
+  }
+
+  private fun String.toMessageAttributeValue(): MessageAttributeValue = MessageAttributeValue()
+    .withDataType("String")
+    .withStringValue(this)
+
+  private fun getDelaySeconds(deliveryDelay: Duration?): Int {
+    return (deliveryDelay?.toMillis() ?: 0 / 1000).toInt()
```

Comment:
@halakaraki what's the precedence order here? Do we want `(deliveryDelay?.toMillis() ?: 0) / 1000` or `(deliveryDelay?.toMillis() ?: (0 / 1000))`?

Reply:
Tangent: why `toMillis()` and then `/1000`? Could we just use `getSeconds()`?

Reply:
working on it